src/interface/pipe.rs

```diff
@@ -89,60 +168,127 @@ impl EmulatedPipe {
             return false;
         }
     }
+
+    /// # Description
+    /// Checks if pipe is currently ready for writing, used by select/poll etc
+    ///
+    /// # Returns
+    ///
+    /// True if descriptor is ready for writing, false if it will block
+    ///
     pub fn check_select_write(&self) -> bool {
         let write_end = self.write_end.lock();
         let pipe_space = write_end.remaining();
 
-        return pipe_space != 0;
+        // Linux considers a pipe writeable if there is at least PAGE_SIZE (PIPE_BUF) remaining space (4096 bytes)
+        return (self.size - pipe_space) > PAGE_SIZE;
```

Comment:
Does this mean there isn't a fully allocated page buffer?  This will not necessarily be exactly the size you mention.

Note also, that I don't know we need to exactly mimic Linux's behavior.  Many of the corner cases just don't matter. in practice.  We can look at what POSIX requires and ensure we're compliant there.

Reply:
I'm not sure I understand your comment but having an empty page in the Linux buffer guarantees there to be at least 4096 bytes? I believe @qianxichen233 mentioned that this is what the posix requriement is.

Reply:
This requirements is from Linux man page, I think it is not a POSIX standard. From what I've read: "The Space for PIPE_BUF bytes column indicates whether the pipe has room to atomically write PIPE_BUF bytes without blocking. This is the criterion on which Linux considers a pipe ready for writing. Some other UNIX implementations use the same criterion; others consider a pipe writable if even a single byte can be written."